src/mc-fle2-payload-iev.c

```diff
@@ -52,6 +55,212 @@ mc_FLE2IndexedEqualityEncryptedValueTokens_new (void)
+      return false;
+   }
+
+   _mongocrypt_buffer_t encryption_out;
+   _mongocrypt_buffer_init (&encryption_out);
+
+   if (!mc_fle2IndexedEncryptedValue_encrypt (crypto,
+                                              ClientEncryptedValue,
+                                              token,
+                                              index_tokens,
+                                              &encryption_out,
+                                              status)) {
+      return false;
+   }
+
+   uint32_t expected_plaintext_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (
+      ClientEncryptedValue->len,
+      (uint32_t) (sizeof (uint64_t) * 2 + sizeof (uint32_t) * 3),
+      &expected_plaintext_size,
+      status));
+
+   uint32_t expected_cipher_size = _mongocrypt_fle2_calculate_ciphertext_len (
+      expected_plaintext_size, status);
+   uint32_t expected_buf_size = 0;
+   CHECK_AND_RETURN (safe_uint32_t_sum (expected_cipher_size,
+                                        (uint32_t) (1 + sizeof (S_KeyId)),
+                                        &expected_buf_size,
+                                        status));
+
+   if (buf->len < expected_buf_size) {
+      CLIENT_ERR (
+         "mc_FLE2IndexedEncryptedValue_write buf is not large enough for iev");
+      return false;
+   }
+
+   mc_writer_t writer;
+   mc_writer_init_from_buffer (&writer, buf, __FUNCTION__);
+
+   const uint8_t subtype =
+      (uint8_t) MC_SUBTYPE_FLE2IndexedEqualityEncryptedValue;
+
+   if ((original_bson_type < 0) || (original_bson_type > 0xFF)) {
+      CLIENT_ERR ("Field 't' must be a valid BSON type, got: %d",
+                  original_bson_type);
+      return false;
+   }
+
+   const uint8_t bson_type = (uint8_t) original_bson_type;
+
+   mc_writer_write_u8 (&writer, &subtype, status);
+   mc_writer_write_buffer (&writer, S_KeyId, S_KeyId->len, status);
+   mc_writer_write_u8 (&writer, &bson_type, status);
+
+   mc_writer_write_buffer (
+      &writer, &encryption_out, encryption_out.len, status);
+
+   _mongocrypt_buffer_cleanup (&encryption_out);
+
+   return true;
+}
+
+#define CHECK_REMAINING_BUFFER_AND_RET(write_size)       \
+   if ((writer->pos + (write_size)) > writer->len) {     \
+      CLIENT_ERR ("%s expected byte "                    \
+                  "length >= %" PRIu64 " got: %" PRIu64, \
+                  writer->parser_name,                   \
+                  writer->pos + (write_size),            \
+                  writer->len);                          \
+      return false;                                      \
+   }
+
+static bool
+mc_fle2IndexedEncryptedValue_encrypt (
+   _mongocrypt_crypto_t *crypto,
+   const _mongocrypt_buffer_t *ClientEncryptedValue,
+   mc_ServerDataEncryptionLevel1Token_t *token,
+   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
+   _mongocrypt_buffer_t *out,
+   mongocrypt_status_t *status)
+{
+   _mongocrypt_buffer_t in;
```

Comment:
`in` and `iv` are leaked.

```
Direct leak of 138 byte(s) in 1 object(s) allocated from:
    #0 0x10ad4d7d0 in wrap_malloc+0xa0 (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x487d0) (BuildId: 7b332288c65134ea8645ebf748e4cb682400000010000000000a0a0000010c00)
    #1 0x10a7a300d in bson_malloc bson-memory.c:70
    #2 0x10a69dcbb in _mongocrypt_buffer_resize mongocrypt-buffer.c:75
    #3 0x10a69df14 in _mongocrypt_buffer_init_size mongocrypt-buffer.c:89
    #4 0x10a67d524 in mc_fle2IndexedEncryptedValue_encrypt mc-fle2-payload-iev.c:221
    #5 0x10a67cbe6 in mc_FLE2IndexedEncryptedValue_write mc-fle2-payload-iev.c:137
    #6 0x10a42fc1e in test_FLE2IndexedEqualityEncryptedValue_decrypt test-mc-fle2-payload-iev.c:323
    #7 0x10a661b77 in main test-mongocrypt.c:1027
    #8 0x7fff20391f3c in start+0x0 (libdyld.dylib:x86_64+0x15f3c) (BuildId: fd8db5bcf19935249dc4daec0e94712f32000000200000000100000000070b00)

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x10ad4d7d0 in wrap_malloc+0xa0 (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x487d0) (BuildId: 7b332288c65134ea8645ebf748e4cb682400000010000000000a0a0000010c00)
    #1 0x10a7a300d in bson_malloc bson-memory.c:70
    #2 0x10a69dcbb in _mongocrypt_buffer_resize mongocrypt-buffer.c:75
    #3 0x10a69df14 in _mongocrypt_buffer_init_size mongocrypt-buffer.c:89
    #4 0x10a67d82b in mc_fle2IndexedEncryptedValue_encrypt mc-fle2-payload-iev.c:255
    #5 0x10a67cbe6 in mc_FLE2IndexedEncryptedValue_write mc-fle2-payload-iev.c:137
    #6 0x10a42fc1e in test_FLE2IndexedEqualityEncryptedValue_decrypt test-mc-fle2-payload-iev.c:323
    #7 0x10a661b77 in main test-mongocrypt.c:1027
    #8 0x7fff20391f3c in start+0x0 (libdyld.dylib:x86_64+0x15f3c) (BuildId: fd8db5bcf19935249dc4daec0e94712f32000000200000000100000000070b00)
```

This reproduces when building with clang 14 on my mac when configuring with `-DCMAKE_C_FLAGS="-fsanitize=address"` with running tests with the environment variable `export ASAN_OPTIONS='detect_leaks=1'`. I will investigate why ASAN tasks in Evergreen are not reporting leaks. 

Here is a suggested alternative, using a `goto` pattern to clean up:

```c
static bool
mc_fle2IndexedEncryptedValue_encrypt (
   _mongocrypt_crypto_t *crypto,
   const _mongocrypt_buffer_t *ClientEncryptedValue,
   mc_ServerDataEncryptionLevel1Token_t *token,
   mc_FLE2IndexedEqualityEncryptedValueTokens *index_tokens,
   _mongocrypt_buffer_t *out,
   mongocrypt_status_t *status)
{
#define CHECK_AND_GOTO(x) \
   if (!(x)) {            \
      goto cleanup;       \
   }

   bool ok = false;
   _mongocrypt_buffer_t in;
   _mongocrypt_buffer_t iv;

   _mongocrypt_buffer_init (&in);
   _mongocrypt_buffer_init_size (&iv, MONGOCRYPT_IV_LEN);

   uint32_t expected_buf_size = 0;
   CHECK_AND_GOTO (
      safe_uint32_t_sum (ClientEncryptedValue->len,
                         (uint32_t) (sizeof (uint64_t) * 2 + (32 * 3)),
                         &expected_buf_size,
                         status));

   _mongocrypt_buffer_resize (&in, expected_buf_size);

   uint32_t ciphertext_len =
      _mongocrypt_fle2_calculate_ciphertext_len (expected_buf_size, status);
   _mongocrypt_buffer_resize (out, ciphertext_len);

   mc_writer_t writer;
   mc_writer_init_from_buffer (&writer, &in, __FUNCTION__);

   uint64_t length;
   length = ClientEncryptedValue->len;
   CHECK_AND_GOTO (mc_writer_write_u64 (&writer, &length, status));

   CHECK_AND_GOTO (mc_writer_write_buffer (
      &writer, ClientEncryptedValue, ClientEncryptedValue->len, status));

   CHECK_AND_GOTO (
      mc_writer_write_u64 (&writer, &index_tokens->counter, status));

   CHECK_AND_GOTO (
      mc_writer_write_prfblock_buffer (&writer, &index_tokens->edc, status));

   CHECK_AND_GOTO (
      mc_writer_write_prfblock_buffer (&writer, &index_tokens->esc, status));

   CHECK_AND_GOTO (
      mc_writer_write_prfblock_buffer (&writer, &index_tokens->ecc, status));

   const _mongocrypt_buffer_t *token_buf =
      mc_ServerDataEncryptionLevel1Token_get (token);

   uint32_t bytes_written;


   if (!_mongocrypt_random (crypto, &iv, MONGOCRYPT_IV_LEN, status)) {
      _mongocrypt_buffer_cleanup (&iv);
      return false;
   }

   CHECK_AND_GOTO (_mongocrypt_fle2_do_encryption (
      crypto, &iv, token_buf, &in, out, &bytes_written, status));

   ok = true;

cleanup:
   _mongocrypt_buffer_cleanup (&iv);
   _mongocrypt_buffer_cleanup (&in);
   return ok;
#undef CHECK_AND_GOTO
}
```